connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1058,14 +1058,20 @@ public void putConnectorConfig(final String connName, final Map<String, String>
                             }
 
                             log.trace("Submitting connector config {} {} {}", connName, allowReplace, configState.connectors());
-                            writeToConfigTopicAsLeader(() -> configBackingStore.putConnectorConfig(connName, config));
-
-                            // Note that we use the updated connector config despite the fact that we don't have an updated
-                            // snapshot yet. The existing task info should still be accurate.
-                            ConnectorInfo info = new ConnectorInfo(connName, config, configState.tasks(connName),
-                                // validateConnectorConfig have checked the existence of CONNECTOR_CLASS_CONFIG
-                                connectorType(config));
-                            callback.onCompletion(null, new Created<>(!exists, info));
+                            Callback<Void> cb = (err, result) -> {
```

Comment:
Another option could be to directly call `callback.onCompletion` in `KafkaConfigBackingStore::putConnectorConfig` itself - i.e. construct the `Created<ConnectorInfo>` object there using its in-memory maps directly rather than the snapshot; this would avoid this double callback kinda mechanism but would need some other refactoring.